lib/app/modules/bulk_data/bulk_data_discovery_sequence.rb

```diff
@@ -0,0 +1,173 @@
+# frozen_string_literal: true
+
+module Inferno
+  module Sequence
+    class BulkDataDiscoverySequence < SequenceBase
+      title 'Bulk Data Discovery'
+
+      test_id_prefix 'Discovery'
+
+      requires :url
+      defines :oauth_authorize_endpoint, :oauth_token_endpoint, :oauth_register_endpoint
+
+      description "Retrieve server's SMART on FHIR configuration"
+
+      details %(
+        # Background
+
+        The #{title} Sequence test looks for authorization endpoints and SMART
+        capabilities as described by the [SMART Backend Services: Authorization
+        Guide](https://build.fhir.org/ig/HL7/bulk-data/authorization/index.html).
+      )
+
+      WELL_KNOWN_FIELDS = [
+        'token_endpoint',
+        'scopes_supported',
+        'token_endpoint_auth_methods_supported',
+        'token_endpoint_auth_signing_alg_values_supported'
+      ].freeze
+
+      SMART_OAUTH_EXTENSION_URL = 'http://fhir-registry.smarthealthit.org/StructureDefinition/oauth-uris'
+
+      REQUIRED_OAUTH_ENDPOINTS = [
+        { url: 'token', description: 'token' }
+      ].freeze
+
+      attr_accessor :well_known_configuration, :conformance
+
+      def oauth2_metadata_from_conformance
+        options = {
+          token_url: nil
+        }
+        begin
+          @conformance.rest.each do |rest|
+            options.merge! @client.get_oauth2_metadata_from_service_definition(rest)
+          end
+        rescue StandardError => e
+          FHIR.logger.error "Failed to locate SMART-on-FHIR OAuth2 Security Extensions: #{e.message}"
+        end
+
+        options.delete_if { |_k, v| v.nil? }
+        options
+      end
+
+      test :read_well_known_endpoint do
+        metadata do
+          id '01'
+          name 'Retrieve Configuration from well-known endpoint'
+          link 'http://www.hl7.org/fhir/smart-app-launch/conformance/#using-well-known'
+          description %(
+            The authorization endpoints accepted by a FHIR resource server can
+            be exposed as a Well-Known Uniform Resource Identifier
+          )
+          optional
+        end
+
+        well_known_configuration_url = @instance.url.chomp('/') + '/.well-known/smart-configuration'
+        well_known_configuration_response = LoggedRestClient.get(well_known_configuration_url)
+        assert_response_ok(well_known_configuration_response)
+        assert_response_content_type(well_known_configuration_response, 'application/json')
+
+        @well_known_configuration = assert_valid_json(well_known_configuration_response.body)
+
+        @well_known_token_url = @well_known_configuration['token_endpoint']
```

Comment:
I think most of these are only used in this test, so there's no reason for them to be instance variables.